### PR TITLE
59 improve classifier

### LINKED ISSUE #59: [JQ] PRIORITY- retraining the classifier
Currently the model uses jobbert embedding layer and then inputs that into logistic regression.

Better options are:

- [x] Fine-tune jobbert (like company description classifier)
- [x] fine-tune smaller classification model (distilbert?)
- [x] Use faster embeddings (all-minilm) and input them into classifier (eg logistic regression again)

Other improvements are:

- [x] training on a balanced sample (previously, it was unbalanced - more 0s than 1s)
- [x] cluster the negative sentences and take a sample from each cluster, so that the negative sample gives good coverage of the type of negative text the algorithm might end up seeing
- [ ] padding out underrepresented dimensions of job quality with fictional examples generated by GPT?
    - [x] label positive samples (in Google Sheets) so that we know if any dimensions are over/underrepresented
    - [x] Find additional examples from underrepresented categories from `prinz-green-jobs/outputs/data/ojo_application/deduplicated_sample/ojo_sample.csv`
    - [x] Label additional examples as positive or negative

Cosmetic/practical improvements include:

- structuring the code in such a way that it's easy to move into ojd_daps_language_models later down the line - eg
    - [x] having hyperparameters in a config
    - [ ] using metaflow

### PR DESCRIPTION
**Pull request by RFOxbury**
---

# Description

Contains scripts for executing hyperparameter sweeps for:

- logistic regression (including a comparison of jobbert embeddings vs all-minilm-l6-v2)
- distilbert
- jobbert

And training scripts with the ideal hyperparameters for:
- distilbert
- jobbert

Plus an inference script for:
- jobbert

It probably isn't necessary to include all of these scripts in `pipeline/` so I could move some to `analysis/`, or delete them before merging?

Fixes #59 

# Instructions for Reviewer

In order to test the code in this PR you need to ...

- [ ] Run `python dap_job_quality/pipeline/sentence_classifier/jobbert_inference.py --n_samples=10` - or set `n_samples` to a higher number, to test how it scales to more data.

Please pay special attention to ...

- Whether the hyperparameters that were tried and then eventually selected make sense.

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: dap_job_quality/notebooks/prepare_final_training_dataset.ipynb [hunk 1]
```diff
@@ -0,0 +1,614 @@
+{
```
Comment (lizgzil) [inline dap_job_quality/notebooks/prepare_final_training_dataset.ipynb:1]: I think the code here needs to be copied out of a notebook and into a script. I think you could do with a subfolder in `pipeline/sentence_classifier/training_data/` that contains this code and the code in `prep_training_data.py`

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py [hunk 2]
```diff
@@ -1,98 +1,318 @@
-import numpy as np
+import ast
```
Comment (lizgzil) [inline dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]: I was a bit confused by what was different about this script and the code in `prepare_final_training_dataset.ipynb`. I see you load the manually labelled output of the notebook in `get_additional_examples_underrepresented_categories`, which has a comment about how it was created.

To make this crystal clear you should add more to the `pipeline/sentence_classifier/README.md`

e.g. a step by step log of what you did and which scripts contain what. 

```
## Training data

To create the training data to train the classifier to predict whether a sentence is about job quality or not, we:
1. Manually labelled job adverts.
2. Ran the notebook `dap_job_quality/notebooks/prepare_final_training_dataset.ipynb`
2. Manually labelled the output of this notebook
3. Concatenate all the training data by running prep_training_data.py which will..


## Training the classifier

To train the classifier we ran:

Note: A Nesta weights and biases account is needed to run this script.

## Inference from the classifier

To make predictions using the classifier run:

python dap_job_quality/pipeline/sentence_classifier/jobbert_inference.py --n_samples=10

```

File: dap_job_quality/notebooks/pad_out_sample.ipynb [hunk 3]
```diff
@@ -0,0 +1,327 @@
+{
```
Comment (lizgzil) [inline dap_job_quality/notebooks/pad_out_sample.ipynb]: It's not clear to me what this notebook is for and where the output of it gets used? If it doesnt get used then may as well delete it.

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py [hunk 4]
```diff
@@ -1,98 +1,318 @@
...
-    # skip the first 10 job ads - we didn't know what we were labelling at that point
-    target_ids = job_ids[10:]
+    additional_data["label"] = additional_data["JQ"].apply(
+        lambda x: 0 if x == "0" or x == "not sure" else 1
+    )
+
+    return additional_data
+
+
+def find_category_representation(additional_data):
```
Comment (lizgzil) [inline dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]:
```suggestion
def find_category_representation(additional_data, positive_sents):
```

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py [hunk 5]
```diff
@@ -1,98 +1,318 @@
...
+    if len(duplicated_ids) < 1:
+        logging.info("IDs are not duplicated")
+    else:
+        raise ValueError(
+            f"IDs are duplicated! Please go back and check the data. The following IDs are duplicated: {duplicated_ids}"
+        )
 
-    labelled_df_clean = filter_job_ads(labelled_df)
+    positive_categories, additional_positives = find_category_representation(
+        additional_data
```
Comment (lizgzil) [inline dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]:
```suggestion
        additional_data, positive_sents
```

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py [hunk 6]
```diff
@@ -1,98 +1,318 @@
...
+    neg_sentences = sentences - sentences_w_spans
+    return neg_sentences
+
+
+def create_neg_sent_df(positive_sents, false_negatives, additional_data):
+    negative_sentences = {}
+
+    for id in positive_sents["id"].unique():
+        neg_sentences = get_negative_example_sentences(positive_sents, id)
+        negative_sentences[id] = list(neg_sentences)
```
Comment (lizgzil) [inline dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]: I think using a groupby will make this a bit cleaner
```suggestion
def get_negative_example_sentences(
    subset, sentences_col="sentences", true_sent_col="sentence"
):
    sentences = set(subset[sentences_col].iloc[0])
    sentences_w_spans = set(subset[true_sent_col].unique())
    neg_sentences = sentences - sentences_w_spans
    return neg_sentences


def create_neg_sent_df(positive_sents, false_negatives, additional_data):
    negative_sentences = {}

    for id, subset in positive_sents.groupby("id"):
        neg_sentences = get_negative_example_sentences(subset)
        negative_sentences[id] = list(neg_sentences)
```

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py [hunk 7]
```diff
@@ -1,98 +1,318 @@
...
+
+    return positive_categories, additional_positives
+
+
+def get_negative_example_sentences(
+    df, id, sentences_col="sentences", true_sent_col="sentence"
+):
+    subset = df[df["id"] == id].copy()
+
+    sentences = set(subset[sentences_col].iloc[0])
```
Comment (lizgzil) [inline dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]: I take it the list of sentences are the same for every row and that why you can just use the first row?

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py [hunk 8]
```diff
@@ -1,98 +1,318 @@
...
+    additional_negatives = additional_negatives.rename(
+        columns={"sentences": "sentence"}
+    )[["id", "sentence"]]
+
+    all_negatives = pd.concat([neg_sent_df_filtered, additional_negatives])
+
+    return all_negatives
+
+
+def sample_negative_sents(all_negatives, n_positives, num_clusters=50):
```
Comment (lizgzil) [inline dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]: this is really nice!

File: requirements.txt [hunk 9]
```diff
@@ -20,3 +20,5 @@ wordcloud
 xlrd
 torch==2.1.2
 tqdm==4.66.1
+umap-learn
+datasets
```
Comment (lizgzil) [inline requirements.txt]: also need altair

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py [hunk 10]
```diff
@@ -1,98 +1,318 @@
...
+    # Create the Altair chart
+    chart = (
+        alt.Chart(all_negatives)
+        .mark_circle(size=60)
+        .encode(x="x:Q", y="y:Q", color="cluster:N", tooltip=["cluster", "sentence"])
+        .properties(width=900, height=600)
+        .interactive()
+    )
+
+    chart.save(PROJECT_DIR / "outputs/figures/negative_sentences_clusters.html")
```
Comment (lizgzil) [inline dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]: looked great to me!

File: dap_job_quality/pipeline/sentence_classifier/log_reg.py [hunk 11]
```diff
@@ -13,6 +13,7 @@
     ConfusionMatrixDisplay,
     accuracy_score,
     f1_score,
+    recall_score,
```
Comment (lizgzil) [inline dap_job_quality/pipeline/sentence_classifier/log_reg.py:26]: what is the difference between this script and whats in `classifier_sweep.py`? if this one is legacy, then delete it

File: dap_job_quality/pipeline/sentence_classifier/jobbert_sweep.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        help="How many runs to do as part of the sweep?",
+    )
+
+    args = parser.parse_args()
+    logging.info(args)
+
+    sweep_id = wandb.sweep(
+        jobbert_config["sweep_config"], project="dap-job-quality", entity="nesta-uk"
+    )
+    wandb.agent(sweep_id, function=model_train, count=args.n_runs)
```
Comment (lizgzil) [inline dap_job_quality/pipeline/sentence_classifier/jobbert_sweep.py:160]: so cool to see how wandb is used!

**Review by lizgzil — COMMENTED** [review]
this looks really great - so much good work! 🎉 

I ran `python dap_job_quality/pipeline/sentence_classifier/jobbert_inference.py --n_samples=10` and the outputs look really reasonable.

I couldn't access W&B so I can't see the results of the parameter sweep / final results though. 

I was thinking it would be nice to store the final training results on s3 somewhere too. I see in `jobbert_train.py` you save something locally `PROJECT_DIR / "outputs/models/sentence_classifier/jobbert-base-cased-jq"` - could you upload this to S3 too?

I don't think there are any bugs/ problems with this PR that I can see! I also ran the `prep_training_data.py` script and it looked good.

The main things that I think are needed before this is merged are:
- Delete legacy files - is pad_out_sample.ipynb used anywhere? also has log_reg.py been replaced by `classifier_sweep.py`?
- Create a clearly structured README explaining what was done, some results + what the scripts are. e.g. how you created the training data (and which scripts were ran), the process of testing different classifiers (and the best results from each in a table), details about the final trained classifier (the results of it and how to use it). This way it would be clear that the jobbert is the sentence classifier used going forward.

Optional:
- Put the scripts which concat and process the training data into their own folder
- Put the prepare_final_training notebook into a script
- my minor code suggestions

File: dap_job_quality/pipeline/sentence_classifier/log_reg.py (hunk 11 above)
Comment (RFOxbury) [reply dap_job_quality/pipeline/sentence_classifier/log_reg.py:26]: Yes it was legacy - deleted now!

File: requirements.txt (hunk 9 above)
Comment (RFOxbury) [reply requirements.txt]: Thanks, added altair and some related dependencies!

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py (hunk 8 above)
Comment (RFOxbury) [reply dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]: Thanks Liz! Your idea :) It worked really well and doing both the visualisation and the clustering in 2d was a good shout I think :)

File: dap_job_quality/notebooks/pad_out_sample.ipynb (hunk 3 above)
Comment (RFOxbury) [reply dap_job_quality/notebooks/pad_out_sample.ipynb]: Thanks, I had to do some digging to remember what this was 😵‍💫 I've refactored it as `pipeline/sentence_classifier/training_data/03_pad_out_sample.py`.

Comment (RFOxbury) [reply dap_job_quality/notebooks/pad_out_sample.ipynb]: *and deleted the notebook

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py (hunk 2 above)
Comment (RFOxbury) [reply dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]: Thanks Liz - I've tried to cover everything to do with data preparation in the readme now!

**Review by lizgzil — APPROVED** [review]
Thanks Rosie! This looks great and so much clearer what happened to create the training data :) 

All good to merge.
